src/runtime/particle-execution-host.ts

```diff
@@ -47,262 +47,301 @@ export class ParticleExecutionHost {
+          console.warn(`'DevToolsConnected()' not supported for Java PEC yet`);
+        }      
+      }(javaPort, arc));
+    }
+  }
 
-      onInitializeProxy(handle: StorageProviderBase, callback: number) {
-        const target = {};
-        handle.on('change', data => this.SimpleCallback(callback, data), target);
-      }
+  get jsApiPort(): PECOuterPort {
+    return this._apiPorts.get('js');
+  }
 
-      async onSynchronizeProxy(handle: StorageProviderBase, callback: number) {
-        const data = await handle.modelForSynchronization();
-        this.SimpleCallback(callback, data);
-      }
+  get javaApiPort(): PECOuterPort {
+    return this._apiPorts.get('java');
+  }
 
-      async onHandleGet(handle: StorageProviderBase, callback: number): Promise<void> {
-        const data = await (handle as SingletonStorageProvider).get();
-        this.SimpleCallback(callback, data);
-      }
+  stop() {
+    this._apiPorts.forEach(apiPort => apiPort.Stop());
+  }
 
-      async onHandleToList(handle: StorageProviderBase, callback: number) {
-        const data = await (handle as CollectionStorageProvider).toList();
-        this.SimpleCallback(callback, data);
-      }
+  get idle(): Promise<Map<Particle, number[]>> | undefined {
+    if (this.idlePromise == undefined) {
+      this.idlePromise = new Promise((resolve, reject) => {
+        this.idleResolve = resolve;
+      });
+    }
+    this.idleVersion = this.nextIdentifier;
+    this._apiPorts.forEach(apiPort => apiPort.AwaitIdle(this.nextIdentifier++));
+    return this.idlePromise;
+  }
 
-      onHandleSet(handle: StorageProviderBase, data: {}, particleId: string, barrier: string) {
-        // TODO: Awaiting this promise causes tests to fail...
-        floatingPromiseToAudit((handle as SingletonStorageProvider).set(data, particleId, barrier));
-      }
+  get messageCount(): number {
+    return [...this._apiPorts.values()].map(apiPort => apiPort.messageCount).reduce((prev, current) => prev + current, 0);
+  }
 
-      onHandleClear(handle: StorageProviderBase, particleId: string, barrier: string) {
-        // TODO: Awaiting this promise causes tests to fail...
-        floatingPromiseToAudit((handle as SingletonStorageProvider).clear(particleId, barrier));
-      }
+  sendEvent(particle, slotName, event): void {
+    this._apiPorts.forEach(apiPort => apiPort.UIEvent(particle, slotName, event));
+  }
 
-      async onHandleStore(handle: StorageProviderBase, callback: number, data: {value: {}, keys: string[]}, particleId: string) {
-        // TODO(shans): fix typing once we have types for Singleton/Collection/etc
-        // tslint:disable-next-line: no-any
-        await (handle as CollectionStorageProvider).store(data.value, data.keys, particleId);
-        this.SimpleCallback(callback, {});
-      }
+  instantiate(particle: Particle, stores: Map<string, StorageProviderBase>): void {
+    const apiPort = particle.isJavaParticle() ? this.javaApiPort : this.jsApiPort;
```

Comment:
store the port in the particle rather than constantly switching like this; that will also set us up better for multiple ports of the same type later.

Reply:
not sure what you mean by "storing IN the particle" - this is the recipe/particle and it is frozen. i added a map from particle to the selected port. can you, please, clarify?

Reply:
OK, yeah a map works well. I mainly didn't want the conditional statement everywhere :)